### ISSUE TITLE
Uploading Folders into shared "upload only" folders flattens hierarchy

### ISSUE THREAD

**Issue by tjohs** [nextcloud/server#15921]
### Steps to reproduce
0. create a folder with content on local drive (e.g. \folder\text.txt, \folder\folder\text.txt)
1. create folder in NextCloud to share
2. share folder externally (i.e. email address)
3. adjust permission to "Upload Only"
4. open link in browser
5. drag&drop folder (step 0) into the browser window
6. files are uploaded (text.txt, text (2).txt)

### Expected behaviour
Within the "shared folder" the dragged folder should be created, keeping the folder/files structure. 

### Actual behaviour
No folder is created. Only the contained files are uploaded. Files within different subfolders will be uploaded into the same "shared folder", folder hierarchy is not kept, files (from subfolders) that are already existing in the "shared folder" will be renamed "filename (2)"... 
This can lead to useless uploads: assuming system backups of similar but not identical systems. They share same configuration file names but different content. If now the hierarchy is flattened while \System1\Conf\config.ini and \System2\Conf\config.ini are uploaded, the "shared folder" will contain "config.ini" and "config (2).ini", but you would not know which file belongs to which system backup... 
This is only one of an endless variation of scenarios.

### Server configuration
**Operating system**: Debian 9
**Web server:** Apache 2
**Database:** mysql
**PHP version:** 7.2.19
**Nextcloud version:** 16.0.1
**Where did you install Nextcloud from:** nextcloudpi
**Are you using encryption:** no

### Client configuration
**Browser:** Firefox 67.0.1 (also tested on Microsoft Edge with similar result)
**Operating system:** Windows 10 / Windows 7

###Conclusion
The option to share folders with colleagues/customers to simply upload files/folders is nice, but the described behaviour leads to useless data. 
An alternative to creating folders as I would have expected would be to automatically create ZIPs upon uploading named as the folder(s)

EDIT: Still the same with NC24.0.6

**Comment by stleitner** [comment]
+1 , normal users do not understand why this is not working as expected. They say: "It just works on webservice XY"

**Comment by despens** [comment]
There is even more weird things going on when uploading a folder with sub-folders in chrome: the sub-folder appears in nextcloud as a zip file if there are conflicting names!

:confused:

**Comment by tjohs** [comment]
> +1 , normal users do not understand why this is not working as expected. They say: "It just works on webservice XY"

:-)  I think does does not only surprise "normal users", but also those running their own servers and thus having more technical background...

**Comment by stefandesu** [comment]
Any update on this? This was an issue for us as well, and I don't understand why this is the case since normal upload (so either on your own account or in a shared folder with "Allow upload and editing") allows uploading a folder structure. It should therefore be an easy fix, right?

**Comment by tjohs** [comment]
Unfortunately not.
Since this issue remains now since several versions of NC, I have sort of given up - although this is a really annoying problem and I cannot understand why this is not considered being a problem ... I agree, it feels like there should be a really easy fix.

**Comment by szaimen** [comment]
cc @nextcloud/sharing does it make sense `keeping the folder/files structure.` for `upload only` folders?

**Comment by despens** [comment]
@szaimen On the most basic level, if a user uploads a folder structure with two files that have the same name but are located in different folders, flattening the structure is causing data loss. (Apart from the fact that flattening file structure itself is data loss.)

**Comment by tjohs** [comment]
> 
> 
> cc @nextcloud/sharing does it make sense `keeping the folder/files structure.` for `upload only` folders?

As @despens commented, it is essential to keep the folder structure not to confuse with the contained files being automatically renamed. One cannot check from which subfolder the files originated from. This is an absolute no-go!

**Comment by tjohs** [comment]
as now already a bot consideres this topic to be marked "stale", I wonder how it is possible to raise the awareness of the developers for this issue? I consider this a real issue that **_must_** be solved!

**Comment by rolfheij** [comment]
Leaving a message here to revive this issue, since also I would very much like to see it resolved

**Comment by Karem-sobhy** [comment]
Hope this resolved soon i also need this and didn't find it when i needed it the most.

**Comment by tjohs** [comment]
unfortunately this issue and bug report does not to be seen... If anyone has an idea how to get this issue to be realized as problematic, please comment!

**Comment by julpec** [comment]
same for me, I think it's a big problem not to keep the hierarchical directory structure. When you work in education and you ask for group work where each group turns in multiple documents, it's really unusable as it is.

**Comment by tjohs** [comment]
I fully agree. For me this is a major issue and I am surprised that this is not being solved since 2.5 years now. I have identified the issue with NC16 already and it is still present in NC24.0.6
Does anyone know, how the developpers can be pointed towards this issue? I had hoped that opening an issue here would be the right place, but obviously this is not the case

**Comment by rolfheij** [comment]
@tjohs Perhaps you can write a post on their [forum](https://help.nextcloud.com/)?

**Comment by tjohs** [comment]
Had done this already... (actually this issue here was based on the suggestion of @JimmyKater )

https://help.nextcloud.com/t/files-drop-uploading-folders/54080/4

**Comment by despens** [comment]
Coming here to let everyone know that uploading folders to a shared file drop directory still does not work in nextcloud 26.0.3. This is really weird, why even use different uploading mechanisms for logged-in and not logged-in users?

**Comment by tjohs** [comment]
This is extremely annoying. 
Does anyone have an idea, how we finally can have this issue being seen by developers - and hopefully solved in the end?

**Comment by despens** [comment]
@tjohs since the issue was [closed by the stale bot](https://github.com/nextcloud/server/issues/15921#event-5257235570) probably the best tactic is to open a new issue?

(People talk about the dangers of AI but a simple stale bot already can cause so much data loss :wink:)

**Comment by labor4** [comment]
i could reconstruct this

**Comment by tjohs** [comment]
let's hope this finally get's solved ;)
thanks @labor4 to have checked this and reopened this issue

**Comment by stefandesu** [comment]
It's weird that the stale bot closes issues even if there are people reacting with a thumbs down on the warning message. Shouldn't that be enough to count as "further activity" on the issue? 

Stale bots shouldn't exist IMO, they only encourage useless "bump" messages so that still relevant issues don't get closed.

**Comment by szaimen** [comment]
Cc @nextcloud/designers @nextcloud/server-frontend

**Comment by marcoambrosini** [comment]
I also think this is a bug. This behaviour is confusing and misleading. If we allow uploading of folders there, folder structure should be preserved.

**Comment by Adenosintriphosphate** [comment]
Just wanted to mention that the hierarchy is also flattened when uploading on android. So maybe this sits deeper?

**Comment by stleitner** [comment]
Any update on this?

**Comment by fizux** [comment]
+1, I use the drop upload feature a lot, and the reported behavior is a significant issue.

**Comment by ReAVRafalSztaba** [comment]
+Still not resolved, it is really a bug that should have been resolved long time ago

**Comment by astrojuanlu** [comment]
hey folks I don't know how the NC team prioritizes issues but please instead of leaving "any updates" comments, upvote the issue by reacting with 👍 on the first comment and optionally subscribe.

This is basic open source netiquette but basically if you don't see any updates, probably there aren't any updates, so no need to ask. There are lots of folks subscribed to this thread so every comment triggers a notification, please avoid that.

Peace and love ❤️

**Comment by susnux** [comment]
@astrojuanlu yes thank you!
Please upvote issues with the 👍 reaction on the initial post so we see what is the most important issue for the community to fix or feature to implement.

**Comment by skjnldsv** [comment]
It's a feature, absolutely.
https://github.com/nextcloud/server/blob/725736a754c81ccee17ba394de4da38f5bfa0e68/apps/dav/lib/Files/Sharing/FilesDropPlugin.php#L60-L62

There's some security reasoning behind this

**Comment by astrojuanlu** [comment]
@skjnldsv Could you point to what exactly is the security reasoning? I skimmed the PR but didn't find anything.

And given that this is in the top 5 % issues of nextcloud/server by upvotes, if this is considered a feature, at the very least a warning or even an error could be shown to the user when this happens?

**Comment by Karem-sobhy** [comment]
5 years and still we don't have closure for this.

**Comment by skjnldsv** [comment]
> @skjnldsv Could you point to what exactly is the security reasoning? I skimmed the PR but didn't find anything.
> 
> And given that this is in the top 5 % issues of nextcloud/server by upvotes, if this is considered a feature, at the very least a warning or even an error could be shown to the user when this happens?

Could have been a feature at some point. But that doesn't mean it have to stay like that :)
Just stating a few contexts because I skimmed through some related issues.

Nonetheless, from what I understood, this was a safety measure to simplify conflicts resolution. 🤔

**Comment by tjohs** [comment]
I seem to be missing some point, but in which universe the current 'solution' can be called a "feature with security reasonning behind". 😉
The code you refer to simply just indicates: 

wharever will be uploaded, remove hierarchy/folder and put all files into root (of the respective folder), rename files if the name already exists 

Might have been the only idea by that time to avoid error or warning messages, simplifying conflicts resolution as commented, but there should be better options today. Removing a folder structure and automatically renaming files without _any_ feedback is simply a no-go in my eyes.

I am still looking forward to a resolution of this issue. Would appreciate, if this finally would be solved.

**Comment by fizux** [comment]
The solution is instead of granting drop-only access to uploaders, the owner grants full edit privileges so that the directory structure can be preserved.  Then any uploader has access to change or delete whatever anyone else uploaded. This is done in the name of enhancing security.

**Comment by despens** [comment]
I'm LOL'ing hard because if this "feature" that is causing data loss was ever intended to "increase security," the workaround is to grant uploaders **full editing rights** when sharing a directory. So secure!! :joy:

**Comment by labor4** [comment]
> renaming files without any feedback is simply a no-go in my eyes

on a serious note:
mitigating the given argument of "no-go": 

preserving filenames indeed prevents headaches, 
in case the structure is intended to be reconnected to a Project DB for instance.

However **renaming the base directory** is IMHO not a serious headache. so
````

/basedir/struc/ture/file.mov
````
becomes

````
/basedir-1/struc/ture/file.mov

````
or even

````
/basedir-20240805_1714/struc/ture/file.mov
````

**Comment by labor4** [comment]
> https://github.com/nextcloud/server/blob/725736a754c81ccee17ba394de4da38f5bfa0e68/apps/dav/lib/Files/Sharing/FilesDropPlugin.php#L60-L62

@skjnldsv I fear a folder upload is just a bunch of nonrelated PUTs. do you know if sabre knows that a folder upload is a bundled request? or any other way this could be packed together?

**Comment by skjnldsv** [comment]
> @skjnldsv I fear a folder upload is just a bunch of nonrelated PUTs. do you know if sabre knows that a folder upload is a bundled request? or any other way this could be packed together?

Are you talking about the bulk upload or chunk upload?
Both of them are NOT supported yet for public upload (but it should be here in NC31! :rocket:)

Otherwise, yeah, they're just multiple PUT requests, without a real link together (as far as the backend knows at least)

**Comment by labor4** [comment]
> > @skjnldsv I fear a folder upload is just a bunch of nonrelated PUTs. do you know if sabre knows that a folder upload is a bundled request? or any other way this could be packed together?
> 
> Are you talking about the bulk upload or chunk upload? Both of them are NOT supported yet for public upload (but it should be here in NC31! 🚀)

I wasn't aware of that

> 
> Otherwise, yeah, they're just multiple PUT requests, without a real link together (as far as the backend knows at least)

So I think a folder upload with put is probably initiated by a js call, and if at all possible that js function would need to send an identificator alongside the PUT, in order for the server to do anything meaningfull (probably a precondition to fix this issue)
Do you agree?
Otherwhise bulk an chunks sounds more elaborate and more managed. Will these implementations supercede the precondition problem we are having here? Meaning will this become a different situation, and to better have a look at this issue later?

**Comment by skjnldsv** [comment]
Thanks for asking!
Not really, the features I listed are completely separate.
Like mentioned a bit above, the flattening was a previous decision that is a mix of simplifying the process and safety measures. Nextcloud 30 implemented a file request feature that allow users to drop into a subfolder, but anything will still be flattened. 
I can raise that discussion for Nextcloud 31, we'll see if the reasonning taken 8 years ago still stands.
As many features will come to the public files views for 31, that might be time to think again about this ticket :)

**Comment by labor4** [comment]
> Thanks for asking! Not really, the features I listed are completely separate. Like mentioned a bit above, the flattening was a previous decision that is a mix of simplifying the process and safety measures. 

I can understand that decision. Some problems just don't come up in the first place if the expectations are lower to begin with. Maybe folder upload wasn't even usual back then, I don't remember.

> Nextcloud 30 implemented a file request feature that allow users to drop into a subfolder, but anything will still be flattened. 

I see, yes lots of variables.
I don't think that flatten will be a problem to change. It's more the conceptual decision behind it and how to avoid conflicts.

If the folder structures are meant to be sacred, and overwrites are forbidden, the prefix needs to be unique (first dir name), so it's either to throw all into a new unique dir by user's pre-decision (maybe even allow overwrites per that one-time directory), or kind of bundle everything automatically, thus teaching the server of the group.

When there is new infrastructure ahead I guess it's best to chill it another round. Lots can change.

> I can raise that discussion for Nextcloud 31, we'll see if the reasonning taken 8 years ago still stands. As many features will come to the public files views for 31, that might be time to think again about this ticket :)

Thanks for the discussion.

**Comment by knofte** [comment]
Yeah, we're also seeing this issue, and we use nextcloud file drop function for storage of financial data.
Or rather tried to use it.
When you have 1200 folders, with same file names inside (due to regulation), and they're all uploaded in root and just changed (1) (2) (3) etc, it becomes a huge headache.

Unfortunately we need the anonymous file drop solution (firewalled), though flattened hierarchy seems to have never made sense to anyone of the users, and I agree. It seems to be one of those relics from 1996 tbh. :)

**Comment by susnux** [comment]
We do the flatten on purpose to prevent users writing into existing folders.
So when implementing this:
- If uploading a folder
  - If the folder already exists, rename the *new* folder to a new name e.g. `folder (1)`.

**Comment by susnux** [comment]
> Why not instead of renaming the original folder (optionally? admin-setting) create a folder for each upload named by date&time of the upload "YYMMDD-HHMMSS"? The complete folder structure could easily be replicated therein... An appended number "YYMMDD-HHMMSS-1" would only be required for the rare cases where different upload requests would receice the identical time stamp otherwise.

I think this changes a lot of current behavior for uploading plain files as you would expect the files to be on the root.

**Comment by tjohs** [comment]
you are certainly right. For this reason I proposed upload-timestamp-folders being an option admins could activate. Otherwise the proposed rule would be applied: upload folder with entire structure, rename new folder to "folder-1" if "folder" already existes.
What do you think?

**Comment by susnux** [comment]
This feature will be much easier to implement once we merged the restructuring of public share DAV in #52182 as then we have proper collections also for public shares.
